reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -1,315 +1,239 @@
+			throw new IllegalArgumentException("maxTaskQueuedPerThread must be strictly positive, was " + maxTaskQueuedPerThread);
 		}
-		this.threadCap = threadCap;
-		this.remainingThreads = threadCap;
-		this.deferredTaskCap = deferredTaskCap;
-		this.remainingDeferredTasks = deferredTaskCap;
-		this.factory = factory;
-		this.idleServicesWithExpiry = new ConcurrentLinkedDeque<>();
-		this.deferredFacades = new ConcurrentLinkedQueue<>();
-		this.allServices = new ConcurrentLinkedQueue<>();
+		this.maxThreads = maxThreads;
+		this.maxTaskQueuedPerThread = maxTaskQueuedPerThread;
+		this.factory = threadFactory;
+		this.clock = Objects.requireNonNull(clock, "A Clock must be provided");
+		this.ttlMillis = ttlMillis;
+
+		this.boundedServices = new BoundedServices(this);
 		this.evictor = Executors.newScheduledThreadPool(1, EVICTOR_FACTORY);
-		this.evictor.scheduleAtFixedRate(() -> this.eviction(System::currentTimeMillis),
-				ttlSeconds,
-				ttlSeconds,
-				TimeUnit.SECONDS);
+		evictor.scheduleAtFixedRate(boundedServices::eviction,
+				ttlMillis,
+				ttlMillis,
+				TimeUnit.MILLISECONDS);
 	}
 
 	/**
-	 * Instantiates the default {@link ScheduledExecutorService} for the BoundedElasticScheduler
-	 * ({@code Executors.newScheduledThreadPoolExecutor} with core and max pool size of 1).
+	 * Create a {@link BoundedElasticScheduler} with the given configuration. Note that backing threads
+	 * (or executors) can be shared by each {@link reactor.core.scheduler.Scheduler.Worker}, so each worker
+	 * can contribute to the task queue size.
+	 *
+	 * @param maxThreads the maximum number of backing threads to spawn, must be strictly positive
+	 * @param maxTaskQueuedPerThread the maximum amount of tasks an executor can queue up
+	 * @param factory the {@link ThreadFactory} to name the backing threads
+	 * @param ttlSeconds the time-to-live (TTL) of idle threads, in seconds
 	 */
-	@Override
-	public ScheduledExecutorService get() {
-		ScheduledThreadPoolExecutor poolExecutor = new ScheduledThreadPoolExecutor(1, factory);
-		poolExecutor.setMaximumPoolSize(1);
-		poolExecutor.setRemoveOnCancelPolicy(true);
-		return poolExecutor;
+	BoundedElasticScheduler(int maxThreads, int maxTaskQueuedPerThread, ThreadFactory factory, int ttlSeconds) {
+		this(maxThreads,
+				maxTaskQueuedPerThread, factory, ttlSeconds * 1000, Clock.tickSeconds(ZoneId.systemDefault()));
 	}
 
-	@Override
-	public void start() {
-		throw new UnsupportedOperationException("Restarting not supported yet");
+	/**
+	 * Instantiates the default {@link ScheduledExecutorService} for the scheduler
+	 * ({@code Executors.newScheduledThreadPoolExecutor} with core and max pool size of 1).
+	 */
+	BoundedScheduledExecutorService createBoundedExecutorService() {
+		return new BoundedScheduledExecutorService(this.maxTaskQueuedPerThread, this.factory);
 	}
 
 	@Override
 	public boolean isDisposed() {
-		return shutdown;
+		return BOUNDED_SERVICES.get(this) == SHUTDOWN;
```

Comment:
Educational question: can't we just refer the (volatile) field here? Why do we need to get it via the updater?

Reply:
for reads, one can use the _volatile_ directly yeah. It is important to do all writes through the field updater as they have slightly relaxed guarantees. I ended up using the field updater everywhere for consistency, as AFAIK the difference is just cosmetic for reads